proposals/007-azure-kms.md

```diff
@@ -0,0 +1,148 @@
+Microsoft Entra China operated by 21Vianet	https://login.partner.microsoftonline.cn
+...
+```
+
+The implication for us is that the oauth endpoint and scope should be configurable. The vault base uri is a natural part of the configuration anyway. We could potentially infer the `scope` from the configured vault base uri.
+
+[National clouds](https://learn.microsoft.com/en-us/entra/identity-platform/authentication-national-cloud)
+
+### TLS
+
+With two remote systems involved we want independently customizable TLS for:
+
+1. the authentication client
+2. the key vault client
+
+So that we can supply custom trust or use insecure testing modes (mock servers using self-signed certs for example).
+
+### DEK Generation
+
+Another difference is that basic Key Vault cannot generate the DEK bytes for us. Managed HSM does have an operation to generate random bytes for us https://learn.microsoft.com/en-us/rest/api/keyvault/?view=rest-keyvault-keys-7.4#key-operations-managed-hsm-only
+
+### Key Identifiers
+
+When we wrap/unwrap we have to identify a key by its name and version. You can obtain the latest version by https://learn.microsoft.com/en-us/rest/api/keyvault/keys/get-key/get-key?view=rest-keyvault-keys-7.4&tabs=HTTP.
+
+The API usually returns it as a `kid` https://learn.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets-certificates#object-identifiers. Like
+
+> For Vaults: https://{vault-name}.vault.azure.net/{object-type}/{object-name}/{object-version}
+
+eg. `https://my-vault.vault.azure.net/keys/my-key/78deebed173b48e48f55abf87ed4cf71`
+
+The Azure SDK then takes advantage of the fact this happens to be the base URL for various operations like wrap/unwrap.
+
+ed. for [wrap key](https://learn.microsoft.com/en-us/rest/api/keyvault/keys/wrap-key/wrap-key?view=rest-keyvault-keys-7.4&tabs=HTTP) the url is `POST {vaultBaseUrl}/keys/{key-name}/{key-version}/wrapkey?api-version=7.4
+`
+
+To minimise EDEK bytes there are a couple of opportunities when encoding the EDEK:
+1. we can exclude the vaultBaseUrl and have the limitation that we work only with a single key vault
+2. we can encode just the keyName and keyVersion, and use that to decrypt
+3. all documented examples of the keyVersion (and from my experimentation too) are 128bits encoded as a hexadecimal string.
+
+So for the EDEK I think we could optimistically try to encode it as `versionByte,keyNameLength,keyName,keyVersionLength,128-bit-decoded,edek`
+and fall back to using the string like `versionByte,keyNameLength,keyName,keyVersionLength,keyVersionString,edek`. So
+we would have either a 16-byte version implying it's the bytes, or a 32 character string.
+
+## Proposed Initial Implementation
+
+1. Support all flavours of Key Vault. The APIs will be the same, just with a different vault base URI.
+2. Support RSA and HSM-RSA key types, wrapping using `RSA-OAEP-256` but emit a warning that it is not quantum-resistant.
+3. Support HSM-AES key type and AES-GCM wrapping
```

Comment:
on this, I think the EDEK should encode which key type we used, so that we can use the appropriate algorithm at key unwrapping time (Which will be 1:1 with whether the key type is RSA or AES)

Reply:
Another thing you can discover up front is whether `unwrapKey` and `wrapKey` are supported by the key, users are able to do things like create an RSA key that doesn't support unwrap/wrap. 

And you can determine whether the key is enabled.